analyzers/src/SonarAnalyzer.CSharp/Rules/InfiniteRecursion_RoslynCfg.cs

```diff
@@ -0,0 +1,224 @@
+            {
+                private readonly BooleanBlockMatcher existReferenceToDeclaringSymbolMatcher;
+
+                public CfgWalkerForMethod(RecursionAnalysisContext<ControlFlowGraph> context)
+                    : base(
+                        context.ControlFlowGraph,
+                        context.AnalyzedSymbol,
+                        context.SemanticModel,
+                        () => context.AnalysisContext.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, context.IssueLocation, "method's recursion")))
+                {
+                    existReferenceToDeclaringSymbolMatcher = new BooleanBlockMatcher(operation =>
+                    {
+                        if (!(operation.Instance.Syntax is InvocationExpressionSyntax invocation))
+                        {
+                            return false;
+                        }
+
+                        return IsInstructionOnThisAndMatchesDeclaringSymbol(invocation.Expression, declaringSymbol, semanticModel);
+                    });
+                }
+
+                protected override bool BlockHasReferenceToDeclaringSymbol(BasicBlock block) =>
+                    existReferenceToDeclaringSymbolMatcher.Matches(block);
+            }
+
+            private class CfgWalkerForProperty : CfgRecursionSearcher
+            {
+                private readonly bool isSet;
+                private readonly BooleanBlockMatcher existReferenceToDeclaringSymbolMatcher;
+
+                public CfgWalkerForProperty(RecursionAnalysisContext<ControlFlowGraph> context, string reportOn, bool isSetAccessor)
+                    : base(
+                        context.ControlFlowGraph,
+                        context.AnalyzedSymbol,
+                        context.SemanticModel,
+                        () => context.AnalysisContext.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, context.IssueLocation, reportOn)))
+                {
+                    isSet = isSetAccessor;
+                    existReferenceToDeclaringSymbolMatcher = new BooleanBlockMatcher(operation =>
+                        operation.Instance.Syntax is var syntax
+                        && TypesForReference.Contains(syntax.GetType())
+                        && MatchesAccessor(syntax)
+                        && IsInstructionOnThisAndMatchesDeclaringSymbol(syntax, declaringSymbol, semanticModel));
+                }
+
+                private static readonly ISet<Type> TypesForReference = new HashSet<Type> { typeof(IdentifierNameSyntax), typeof(MemberAccessExpressionSyntax) };
+
+                protected override bool BlockHasReferenceToDeclaringSymbol(BasicBlock block) =>
+                    existReferenceToDeclaringSymbolMatcher.Matches(block);
+
+                private bool MatchesAccessor(SyntaxNode node)
+                {
+                    var expr = (ExpressionSyntax)node;
+                    var propertyAccess = expr.GetSelfOrTopParenthesizedExpression();
```

Comment:
I'd collapse them, we don't need to store `expr`.
```suggestion
                    var propertyAccess = ((ExpressionSyntax)node).GetSelfOrTopParenthesizedExpression();
```